ash/src/extensions/ext/drm_format_modifier.rs

```diff
@@ -0,0 +1,83 @@
+use crate::prelude::*;
+use crate::vk;
+use crate::{Device, Instance};
+use std::ffi::CStr;
+use std::mem;
+
+#[derive(Clone)]
+pub struct DrmFormatModifier {
+    handle: vk::Device,
+    fp: vk::ExtImageDrmFormatModifierFn,
+}
+
+impl DrmFormatModifier {
+    pub fn new(instance: &Instance, device: &Device) -> Self {
+        let handle = device.handle();
+        let fp = vk::ExtImageDrmFormatModifierFn::load(|name| unsafe {
+            mem::transmute(instance.get_device_proc_addr(handle, name.as_ptr()))
+        });
+        Self { handle, fp }
+    }
+
+    /// <https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/vkGetImageDrmFormatModifierPropertiesEXT.html>
+    pub unsafe fn get_image_drm_format_modifier_properties(
+        &self,
+        image: &vk::Image,
+        properties: &mut vk::ImageDrmFormatModifierPropertiesEXT,
+    ) -> VkResult<()> {
+        (self.fp.get_image_drm_format_modifier_properties_ext)(self.handle, *image, properties)
+            .result()
+    }
+
+    pub unsafe fn get_drm_format_properties_list(
+        instance: &Instance,
+        pdevice: vk::PhysicalDevice,
+        format: vk::Format,
+        format_properties: vk::FormatProperties,
+    ) -> Vec<vk::DrmFormatModifierPropertiesEXT> {
+        let mut list = vk::DrmFormatModifierPropertiesListEXT::default();
+
+        // Get count of DRM format modifiers for the format.
+        {
+            let mut format_properties_2 = vk::FormatProperties2::builder()
+                .format_properties(format_properties)
+                .push_next(&mut list);
+
+            instance.get_physical_device_format_properties2(
+                pdevice,
+                format,
+                &mut format_properties_2,
+            );
+        }
+
+        let mut data = Vec::with_capacity(list.drm_format_modifier_count as usize);
+        list.p_drm_format_modifier_properties = data.as_mut_ptr();
```

Comment:
Note that we have helper functions for this pattern. It may be a little less trivial to use thanks to the extra struct holding a count and pointer, but it should suffice.

Reply:
This might actually be quite a challenge to use the helper functions:

1. Since `DrmFormatModifierPropertiesListEXT` is mutably borrowed inside `FormatProperties2`, we end up double borrowing when trying to set pointer to the vec inside `DrmFormatModifierPropertiesListEXT` during the second roundtrip after the function has given us the number of drm format modifier properties structs the implementation has to give us.

2. Even if we make the following change to `read_into_uninitialized_vector`, the double borrow issue is still there.:
```diff
pub(crate) unsafe fn read_into_uninitialized_vector<N: Copy + Default + TryInto<usize>, T>(
-    f: impl Fn(&mut N, *mut T) -> vk::Result,
+    mut f: impl FnMut(&mut N, *mut T) -> vk::Result,
) -> VkResult<Vec<T>>
```